### PR TITLE
4-RML-Star-Mappings

### PR DESCRIPTION
**Pull request by jhculb**
Not ready for merge.

This mapping would I think parse based on paper374. 
This also feels like a manual effort to define these, rather than an automated one.
I need to check the typing (e.g. whether rml:TriplesMap or rml:NonAssertedTriplesMap, or other datatypes I haven't found yet are more appropriate.)
I also need to find a guide to parsing (I am confident for example the URI :rr and :rml are not written in this format)

### CODE REVIEW

**Comment by jhculb** [comment]
Todo: Find correct URI for rml prefix

**Review by dachafra — APPROVED** [review]

**Comment by dachafra** [comment]
RDF is `http://www.w3.org/1999/02/22-rdf-syntax-ns#` ;-)

**Comment by jhculb** [comment]
Ready for review and then merge if no comments @andimou @dachafra

File: src/rml-star-mappings/syntax-2.rml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        ]
+    ] .
+
+:secondTM a rml:TriplesMap ;
+    rml:logicalSource [
+        rml:source "data.csv";
+        rml:referenceFormulation ql:CSV
+    ];
+    rml:subjectMap [
+            rml:quotedTriplesMap :firstTM
```
Comment (dachafra) [inline src/rml-star-mappings/syntax-2.rml:30]: this is not correct, the quotedTriplesMap should be in the object (line 20), and here you generated a `rml:NonAssertedTriplesMap` with "s,p,o" values

File: src/rml-star-mappings/syntax-3.rml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            rml:quotedTriplesMap :firstTM
+    ];
+    rr:predicateObjectMap [
+        rr:predicate ex:q; ;
+        rml:objectMap [
+            rml:quotedTriplesMap :secondTM
+        ]
+    ] .
+
+:secondTM a rml:TriplesMap ;
```
Comment (dachafra) [inline src/rml-star-mappings/syntax-3.rml:39]: be careful, this is thirdTM not second

File: src/rml-star-mappings/syntax-3.rml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+@prefix rr: <http://www.w3.org/ns/r2rml#> .
+@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix ql: <http://semweb.mmlab.be/ns/ql#> .
+@prefix ex: <http://example/> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#integer> .
+@prefix : <http://example.org> .
+@base <http://example.org> .
+:firstTM a rml:TriplesMap ;
```
Comment (dachafra) [inline src/rml-star-mappings/syntax-3.rml:9]: here firstTM and thirdTM are as well `rml:NonAssertedTriplesMap`
see:
image

File: src/rml-star-mappings/nested-1.rml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        rml:source "data.csv";
+        rml:referenceFormulation ql:CSV
+    ];
+    rml:subjectMap [
+        rml:quotedTriplesMap :secondTM
+    ];
+    rr:predicateObjectMap [
+        rr:predicate ex:q ;
+        rml:objectMap [
+            rml:reference "c7"
```
Comment (dachafra) [inline src/rml-star-mappings/nested-1.rml:50]: semicolon at the end?

**Review by dachafra — CHANGES_REQUESTED** [review]
Review the concept of "NonAssertedTriplesMap", the rest is perfect!

File: src/rml-star-mappings/nested-1.rml (hunk 4 above)
Comment (jhculb) [reply src/rml-star-mappings/nested-1.rml:50]: Taken care of in branch 6-

File: src/rml-star-mappings/syntax-3.rml (hunk 3 above)
Comment (jhculb) [reply src/rml-star-mappings/syntax-3.rml:9]: Ah I think I see: the csv is saying s1 p1 o1, and we're quoting it. For TriplesMaps we are following the ontology and saying "these triples are linked"?

Regardless, to fix (and I'll copy through all the rest) where it is "s, p, o" and not "<another triple> p o" (or rearrangement) I'll use NonAssertedTriplesMap. Is that accurate?

File: src/rml-star-mappings/syntax-3.rml (hunk 2 above)
Comment (jhculb) [reply src/rml-star-mappings/syntax-3.rml:39]: Nice spot, thank you!

File: src/rml-star-mappings/syntax-2.rml (hunk 1 above)
Comment (jhculb) [reply src/rml-star-mappings/syntax-2.rml:30]: Yes, sloppy work here. Thanks for the spot!

File: src/rml-star-mappings/syntax-3.rml (hunk 3 above)
Comment (jhculb) [reply src/rml-star-mappings/syntax-3.rml:9]: Ah that's clear now, thank you

Comment (jhculb) [reply src/rml-star-mappings/syntax-3.rml:9]: And to confirm, << <s> <p> <o> >> <p> <q>
The x=<<<s><p><o>>> triple is nonAsserted
But x <p> <q> is just a triplesMap

File: src/rml-star-mappings/syntax-5.ttl [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        rr:template "http://example/{c1}"
+    ];
+    rr:predicateObjectMap [
+        rr:predicate ex:p1 ;
+        rml:objectMap [
+            rr:template "http://example/{c3}"
+        ]
+    ] .
+
+:firstJoinTM a rml:TriplesMap ;
```
Comment (dachafra) [inline src/rml-star-mappings/syntax-5.ttl:24]: I think this is also NonAsserted, right?

File: src/rml-star-mappings/syntax-5.ttl [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        rr:template "http://example/{c9}"
+    ];
+    rr:predicateObjectMap [
+        rr:predicate ex:p3 ;
+        rml:objectMap [
+            rr:template "http://example/{c11}"
+        ]
+    ] .
+
+:secondJoinTM a rml:TriplesMap ;
```
Comment (dachafra) [inline src/rml-star-mappings/syntax-5.ttl:84]: I think this is also NonAsserted, right?

**Review by dachafra — CHANGES_REQUESTED** [review]
Review syntax5 I think that "internal-join" TriplesMap should be NonAsserted as well
image

File: src/rml-star-mappings/syntax-5.ttl (hunk 6 above)
Comment (jhculb) [reply src/rml-star-mappings/syntax-5.ttl:84]: Yes to both, the only assertion is the one with the q2 predicate :) Fixing now

**Comment by jhculb** [comment]
> Review syntax5 I think that "internal-join" TriplesMap should be NonAsserted as well image

Thanks for the spot!

**Pull request merged**